extended/src/main/java/apoc/ml/Prompt.java

```diff
@@ -0,0 +1,192 @@
+            } catch (QueryExecutionException quee) {
+                if (log.isDebugEnabled())
+                    log.debug("Generated query for question %s\n%s\nfailed with %s".formatted(question, query, quee.getMessage()));
+                retries--;
+                if (retries <= 0) throw quee;
+            }
+        } while (true);
+    }
+
+    @Procedure
+    public Stream<StringResult> schema(@Name(value = "apiKey", defaultValue = "null") String apiKey) throws MalformedURLException, JsonProcessingException {
+        String schemaExplanation = prompt("Please explain the graph database schema to me and relate it to well known concepts and domains.",
+                EXPLAIN_SCHEMA_PROMPT, "This database schema ", loadSchema(tx), apiKey);
+        return Stream.of(new StringResult(schemaExplanation));
+    }
+
+    @Procedure(mode = Mode.READ)
+    public Stream<QueryResult> cypher(@Name("question") String question, @Name(value = "count", defaultValue = "1") Long count,
+                                      @Name(value = "apiKey", defaultValue = "null") String apiKey) {
+        String schema = loadSchema(tx);
+        return LongStream.rangeClosed(1, count).mapToObj(i -> tryQuery(question, apiKey, schema));
+    }
+
+    @NotNull
+    private QueryResult tryQuery(String question, String apiKey, String schema) {
+        String query = "";
+        try {
+            query = prompt(question, SYSTEM_PROMPT, "Cypher Statement (in backticks):", schema, apiKey);
+            // doesn't work right now, fails with security context error
+            // tx.execute("EXPLAIN " + query).close(); // TODO query plan / estimated rows?
+            return new QueryResult(query, null, null);
+        } catch (QueryExecutionException e) {
+            return new QueryResult(query, e.getMessage(), e.getStatusCode());
+        } catch (Exception e) {
+            return new QueryResult(query, e.getMessage(), e.getClass().getSimpleName());
+        }
+    }
+
+    @NotNull
+    private String prompt(String userQuestion, String systemPrompt, String assistantPrompt, String schema, String apiKey) throws JsonProcessingException, MalformedURLException {
+        List<Map<String, String>> prompt = new ArrayList<>();
+        if (systemPrompt != null && !systemPrompt.isBlank()) prompt.add(Map.of("role", "system", "content", systemPrompt));
+        if (schema != null && !schema.isBlank()) prompt.add(Map.of("role", "system", "content", "The graph database schema consists of these elements\n" + schema));
+        if (userQuestion != null && !userQuestion.isBlank()) prompt.add(Map.of("role", "user", "content", userQuestion));
+        if (assistantPrompt != null && !assistantPrompt.isBlank()) prompt.add(Map.of("role", "assistant", "content", assistantPrompt));
+        String result = OpenAI.executeRequest(apiKey, Map.of(), "chat/completions",
+                        "gpt-3.5-turbo", "messages", prompt, "$", apocConfig)
```

Comment:
the model is hardcoded, and as GPT-4 is now in GA maybe we can think about having it as a configuration parameter, as we did in `OpenAI` class